cmd/cansnoop.8
.\" $Id: cansnoop.8,v 1.2 2001/07/27 10:24:55 garlick Exp $
.\"
.TH CANSNOOP 8 "14 May 1999"
.SH NAME
cansnoop \- watch traffic on the Meiko CS/2 CAN bus
.SH SYNOPSIS
.B cansnoop
.RB [-p]
.RB [-h]
.SH DESCRIPTION
.I cansnoop
displays all packets received by the local CAN chip.  With the -p option,
.I cansnoop
instructs the device driver to enter promiscuous mode and receive all packets
present on the LCAN.  With the -h option, heartbeat traffic is not displayed.
.LP
Output looks like this:
.LP
.nf
0.000 011->004  RO  m1       HEARTBEAT     00 00 00 00 ....
2.310 011->004  RO  m1       HEARTBEAT     00 00 00 00 ....
4.610 011->004  RO  m1       HEARTBEAT     00 00 00 00 ....
.fi
.LP
Column 1: the number of seconds elapsed since 
.I cansnoop
was started.
.LP
Column 2: source and destination LCAN address
.LP
Column 3: query type
.LP
Columns 4-5:  extended CAN address and object ID (extended can address is 
mapped to a hostname if present in /etc/canhosts).
.LP
Columns 6-10 up to four bytes of data in hex and ASCI form.
.SH AUTHORS
Jim Garlick <garlick@llnl.gov>
.SH SEE ALSO
canctrl(8), cancon(8), canping(8), can(4)


cmd/ChangeLog
Wed Jul 25 19:21:43 PDT 2001 garlick@llnl.gov

	* Added README, canobj, canhosts.template, DISCLAIMER
	* Created cs2.spec file.
	* Tested under kernel-2.2.19-6.2.7cs2.2

	* Updated bargraph demo for new mmap/ioctl methods (leddemo.c)
	
	* Added elanclock.c

	* Updated spec file for release 0.1-3

Fri Jul 27 04:16:20 PDT 2001 garlick@llnl.gov

	* Make canping send/ack synchronous with timeout/retry (canping.c)
	* Added -f (flood) and -c (count) options (canping.c)
	* Use elanclock instead of gettimeofday for measurement (canping.c)
	* Return value indicates whether target responded (canping.c)

	* Added TESTRW object to decode list (cansnoop.c)

	* Added can_getobjbyid() and can_getobjbyname() (canlib.[c,h], canobj.h)
	* Added can_str2type() and can_type2str() (canlib.[c,h])
	* Added can_hb2str() and can_str2hb() (canlib.[c,h])
	* Renamed canlib.[c,h] to can.[c,h], create libcan.a (all)
	* Use above functions (all)

	* Update for can_dat datatype and new struct alignment (all)

	* Updated spec file for release 0.1-4

Fri Aug  3 11:12:27 PDT 2001

	* Moved leddemo to testled, elanclock to testclock and don't install
	* check gettimeofday and elan nsec clock accuracy (elanclock.c)


cmd/cansnoop.c
/*****************************************************************************\
 *  Copyright (c) 2000 Regents of the University of California
 *  the Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 *  UCRL-CODE-2000-010 All rights reserved.
 *
 *  This file is part of the M/Linux linux port to Meiko CS/2.
 *  For details, see https://github.com/garlick/meiko-cs2
 *
 *  This program is free software; you can redistribute it and/or modify it
 *  under the terms of the GNU General Public License as published by the Free
 *  Software Foundation; either version 2 of the license, or (at your option)
 *  any later version.
 *
 *  Flux is distributed in the hope that it will be useful, but WITHOUT
 *  ANY WARRANTY; without even the IMPLIED WARRANTY OF MERCHANTABILITY or
 *  FITNESS FOR A PARTICULAR PURPOSE.  See the terms and conditions of the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 *  See also:  http://www.gnu.org/licenses/
\*****************************************************************************/

/*
 * $Id: cansnoop.c,v 1.7 2001/07/31 08:53:59 garlick Exp $
 */

#include <stdio.h>
#include <unistd.h>
#include <sys/fcntl.h>
#include <sys/ioctl.h>
#include <assert.h>
#include <asm/param.h>	/* for HZ */
#include <stdint.h>	/* uintN_t types */
#include "can.h"

#define PKTSIZE		(sizeof(struct can_packet))

static char *
obj_to_str(int type, int id)
{
	static struct canobj obj;
	static char numericName[MAXHOSTNAMELEN];

	sprintf(numericName, "%4.4x", id);
	if (type == CANTYPE_ACK || type == CANTYPE_NAK || type == CANTYPE_DAT)
		return numericName;
	if (can_getobjbyid(id, &obj) < 0)
		return numericName;
	return obj.name;
}

#define isprint(c)	((c) >= 0x20 && (c) <= 0x7e)
#define CHR(c)		(isprint(c) ? (c) : '.')

/*
 * 6.477 013->01d  ACK 00,00,13 0031        00 00 00 04 '....'
 * 6.555 013->00c  RO  00,00,0c HEARTBEAT   00 00 00 00 '....'
 * 6.555 00c->013  ACK 00,00,0c 0031        00 00 00 04 '....'
 */
static void 
decode(struct can_packet *pkt, int no_heartbeat)
{
	static unsigned long last_stamp = 0;
	static unsigned long first_stamp = 0;
	static float sec_stamp;
	struct canhostname ch;
	char tmp1[255] = "";
	char tmp2[255] = "";
	int i;

	if (no_heartbeat && pkt->ext.ext.object == CANOBJ_HEARTBEAT)
		return;

	if (first_stamp == 0)
		first_stamp = pkt->timestamp;
	sec_stamp = (float)(pkt->timestamp - first_stamp) / HZ;

	/* can header */
	printf("%-.3f %3.3x->%3.3x  ", 
	    sec_stamp, pkt->can.can.src, pkt->can.can.dest);
	if (pkt->can.can.length < 4) { 
		printf("len=%d - ERROR bad packet length\n", 
		    pkt->can.can.length);
		return;
	}

	assert(pkt->timestamp >= last_stamp);
	last_stamp = pkt->timestamp;

	/* extended header */
	printf("%-3.3s ", can_type2str(pkt->ext.ext.type));

	if (can_gethostbyaddr(pkt->ext.ext.cluster, pkt->ext.ext.module,
	    pkt->ext.ext.node, &ch) == -1)
		printf("%2.2x,%2.2x,%2.2x ", 
		    pkt->ext.ext.cluster,pkt->ext.ext.module,pkt->ext.ext.node);
	else
		printf("%-8.8s ", ch.hostname);

	printf("%-12.12s", obj_to_str(pkt->ext.ext.type, pkt->ext.ext.object));
		
	/* 0-4 bytes of data */
	for (i = 0; i < pkt->can.can.length - 4; i++) {
		sprintf(tmp1 + strlen(tmp1), "%2.2x ", pkt->dat.dat_b[i]);
		sprintf(tmp2 + strlen(tmp2), "%c", CHR(pkt->dat.dat_b[i]));
	}
	printf("  %-11.11s %-4.4s\n", tmp1, tmp2);
}

#define NPKT	1

int
main(int argc, char *argv[])
{
	struct can_packet pkt[NPKT];	
	int fd, i, bytes, packets; 
	int no_heartbeat = 0;

#if 0
	printf("CAN Packets are size %d\n", PKTSIZE);
	printf("timestamp %d\n", (unsigned long)&pkt[0].can.can - (unsigned long)&pkt[0].timestamp);
	printf("can %d\n", (unsigned long)&pkt[0].ext - (unsigned long)&pkt[0].can.can);
	printf("ext %d\n", (unsigned long)&pkt[0].data - (unsigned long)&pkt[0].ext.ext);
#endif

	fd = open("/dev/can", O_RDONLY);
	if (fd < 0) {
		perror("/dev/can");
		exit(1);
	}
	/* request to receive packets sent by this node to someone else */
	if (ioctl(fd, CAN_SET_SNOOPY) < 0) {
		perror("ioctl");
		exit(1);
	}
	while (argc > 1) {
		if (!strcmp(argv[1], "-p")) {
			if (ioctl(fd, CAN_SET_PROMISCUOUS) < 0)
				perror("ioctl");
		} else if (!strcmp(argv[1], "-h")) {
			no_heartbeat = 1;
		} else {
			fprintf(stderr, "Usage: cansnoop [-p] [-h]\n");
			exit(1);
		}
		argc--;
		argv++;
	}

	do {
		bytes = read(fd, pkt, PKTSIZE * NPKT);
		packets = bytes / PKTSIZE;
		assert(bytes % PKTSIZE == 0);

		for (i = 0; i < packets; i++)
			decode(&pkt[i], no_heartbeat);
	} while (packets >= 1);

	close(fd);
	exit(0);
}
